pkg/pillar/cmd/README.md

```diff
@@ -0,0 +1,273 @@
+# Agent Services
+
+This directory contains agent services for pillar.
+
+Each subdirectory contains exactly one agent, whose name is the name of the directory.
+
+Look at [verifier](./verifier/) for an example of the implementation.
+
+## Design Pattern
+
+The design pattern for each agent is as follows. When creating a new agent, ensure
+that you follow the design pattern.
+
+The package name in the agent's directory is the same as the directory name.
+It should contain at least the following files:
+
+* `README.md` describing the agent and its purpose.
+* `run.go` containing the main library entry point for the agent.
+
+In addition, it should contain the following subdirectories:
+
+* `lib/` contains core logic, has no dependency on being called command-line or pubsub, unless pubsub is part of the core business logic. It is important to keep these functions as simple and straightforward as possible. Unit tests of business logic should be here as well.
+* `pubsub/` contains functions to handle pubsub messages, including all handlers. There should be no business logic here, only pubsub message handling. For business logic, handlers should call business logic functions in `lib/`.
```

Comment:
For things like the metadata cmd, where all the business logic is to pass between pubsub and a http server, what do we recommend?

 Put all of it in lib/, all in pubsub/, or rework so that the business logic calls to pubsub Get and Publish are replaced by calling some internal functions which can be stubbed out for testing?

FWIW same question will apply to pubsub-heaving microservices in zedmanager.

Reply:
This is a fair point. I don't have an obvious answer. My immediate one is, this is iterative. My intent is, once this works, we extend it to the bulk of other services which do not use pubsub as core business logic, but rather as a messaging system. Once those are done, we can attack those.

Or we can ask @uncleDecart to come up with a good solution, since he is doing the independent metadata server. 😁